Add RenderMan renderer backend

johnhaddon opened the pull request:
This builds on top of the recent IECoreRenderMan PRs by adding a functioning renderer backend for RenderMan, along with a display driver and unit tests. None of this is exposed in Gaffer yet - although you could import IECoreRenderMan and render to RenderMan, it won't be much use until a future PR provides nodes for building RenderMan shader networks and settings options and attributes.

It's unavoidable that there's quite a lot of code here, but I've tried to split the commits down component-by-component in an order that guides you through things such that at least each component arrives after the others it depends on.

There are plenty of todos in the code, but at this point I consider them much lower priority to making progress on as-yet untouched features. Likewise I haven't yet done the licensing dance needed to run the unit tests in CI, although you should be able to run them locally to satisfy yourself that things are working as advertised. Unless you're on Windows, since I haven't done the env setup work for that yet. I'll deal with that and the CI testing in a future PR.

johnhaddon commented:
The fixups were piling up on this one, so with @murraystevenson's permission I've pushed a version which squashes them all in, with some additional fixes for a lighting bug that crept in recently.

murraystevenson commented on python/IECoreRenderManTest/RendererTest.py:
```diff
@@ -0,0 +1,1381 @@
+			renderer.attributes( IECore.CompoundObject( {
+				"ri:light" : IECoreScene.ShaderNetwork(
+					shaders = {
+						"output" : IECoreScene.Shader( "PxrPortalLight", "ri:light", { "exposure" : 4.0 } ),
```
Is the exposure parameter intended here? Might also be worth adding a test for the portal correctly affecting the intensity and colour of the dome light?

johnhaddon replied:
Looks like that was a copy-paste error - portal lights don't have their own exposure. Corrected that and added the suggested test in https://github.com/GafferHQ/gaffer/pull/6271/commits/c3a60d6fe404c693b01948636ee1b437fd4d4510.

murraystevenson commented on src/IECoreRenderMan/MaterialCache.cpp:
```diff
@@ -0,0 +1,110 @@
+{
+}
+
+ConstMaterialPtr MaterialCache::getMaterial( const IECoreScene::ShaderNetwork *network )
```
I found the slight differences between this and `getDisplacement()` a little confusing, maybe I'm missing the subtleties but shouldn't they basically do the same thing with different caches?

johnhaddon replied:
Indeed. There were more differences at one point in development, with the MaterialCache implementing a fallback to a facing ratio shader for null materials. I moved that to Attributes, but in doing so left behind some unnecessary differences with the displacement code. I've brought the two into alignment in https://github.com/GafferHQ/gaffer/pull/6271/commits/12d0a237196b15e92f3f84f83a2978a992ef8ba0.

murraystevenson commented on src/IECoreRenderMan/Renderer.cpp:
```diff
@@ -0,0 +1,254 @@
+		IECore::DataPtr command( const IECore::InternedString name, const IECore::CompoundDataMap &parameters ) override
+		{
+			const IECore::MessageHandler::Scope messageScope( m_messageHandler.get() );
+			return nullptr;
```
Even though we don't handle any commands here, is it worth warning about unknown commands like we do for the other renderer backends?

johnhaddon replied:
Good idea - done in https://github.com/GafferHQ/gaffer/pull/6271/commits/c03de0b16f0894570ebe73bcfbfa402e7f58d5aa.

murraystevenson commented on src/IECoreRenderMan/Session.cpp:
```diff
@@ -0,0 +1,457 @@
+//
+// I don't really know why this is, but I assume that somehow the name
+// is used to share an acceleration table or some such behind the scenes.
+// Why it is should be our responsibility to facilitate that is beyond me.
```
"Why it should be our"

murraystevenson reviewed (commented):
Thanks John! There is quite a lot here, but I do appreciate you taking the time to split things out into logical commits, it definitely makes things much easier to follow. I do like the structure of all this, with some of the more _interesting_ corners of the Riley API wrapped up as they are...

I've a few minor comments inline but overall would be happy to see this merged so we can move on to the next round of PRs and fix anything that crops up as we go.

johnhaddon commented:
> I've a few minor comments inline but overall would be happy to see this merged so we can move on to the next round of PRs and fix anything that crops up as we go.

Thanks - I've addressed the comments, and am merging so I can get the yet-to-be-reviewed parts rebased for further development.
